receiver/cloudflarereceiver/logs.go

```diff
@@ -0,0 +1,270 @@
+	l.wg.Add(1)
+	defer l.wg.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-l.doneChan:
+			return
+		case <-t.C:
+			err := l.poll(ctx)
+			if err != nil {
+				l.logger.Error("there was an error during the poll", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (l *logsReceiver) poll(ctx context.Context) error {
+	startTime := l.nextStartTime
+	if l.record != nil {
+		startTime = l.record.LastRecordedTime.Format(time.RFC3339)
+		l.record = nil
+	}
+	endTime := time.Now().Format(time.RFC3339)
+
+	err := l.pollForLogs(ctx, startTime, endTime)
+	if err != nil {
+		return err
+	}
+
+	l.nextStartTime = endTime
+	return l.writeCheckpoint(ctx)
+}
+
+func (l *logsReceiver) pollForLogs(ctx context.Context, startTime, endTime string) error {
+	resp, err := l.client.MakeRequest(ctx, startTime, endTime)
+	if err != nil {
+		l.logger.Error("unable to retrieve logs from Cloudflare", zap.Error(err))
+		return err
+	}
+
+	observedTime := pcommon.NewTimestampFromTime(time.Now())
+	logs := l.processLogs(observedTime, resp)
+	if logs.LogRecordCount() > 0 {
+		if err = l.consumer.ConsumeLogs(ctx, logs); err != nil {
+			l.logger.Error("unable to consume logs", zap.Error(err))
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (l *logsReceiver) processLogs(now pcommon.Timestamp, logs []*models.Log) plog.Logs {
+	pLogs := plog.NewLogs()
+
+	for _, log := range logs {
+		resourceLogs := pLogs.ResourceLogs().AppendEmpty()
+		logRecord := resourceLogs.ScopeLogs().AppendEmpty().LogRecords().AppendEmpty()
+		resourceAttributes := resourceLogs.Resource().Attributes()
+		attrs := logRecord.Attributes()
+
+		logRecord.SetObservedTimestamp(now)
+		resourceAttributes.PutStr("cloudflare.ZoneID", l.cfg.Zone)
+
+		ts := time.Unix(log.EdgeEndTimestamp, 0)
+		logRecord.SetTimestamp(pcommon.NewTimestampFromTime(ts))
+		logRecord.SetSeverityNumber(severityFromAPI(log.EdgeResponseStatus))
+		logRecord.SetSeverityText(severityFromAPI(log.EdgeResponseStatus).String())
+
+		bodyBytes, err := json.Marshal(logs)
+		if err != nil {
+			l.logger.Warn("unable to marshal log into a body string")
+			continue
+		}
+		logRecord.Body().SetStr(string(bodyBytes))
+
+		// These attributes are optional and may not be present, depending on the selected fields.
+		putStringToMapNotNil(attrs, "ClientIP", &log.ClientIP)
+		putStringToMapNotNil(attrs, "ClientRequestHost", &log.ClientRequestHost)
+		putStringToMapNotNil(attrs, "ClientRequestMethod", &log.ClientRequestMethod)
+		putStringToMapNotNil(attrs, "ClientRequestURI", &log.ClientRequestURI)
+		putIntToMapNotNil(attrs, "EdgeEndTimestamp", &log.EdgeEndTimestamp)
+		putIntToMapNotNil(attrs, "EdgeResponseBytes", &log.EdgeResponseBytes)
+		putIntToMapNotNil(attrs, "EdgeResponseStatus", &log.EdgeResponseStatus)
+		putIntToMapNotNil(attrs, "EdgeStartTimestamp", &log.EdgeStartTimestamp)
+		putStringToMapNotNil(attrs, "RayID", &log.RayID)
+	}
+
+	return pLogs
+}
+
+func putStringToMapNotNil(m pcommon.Map, k string, v *string) {
+	if v != nil {
```

Comment:
Do you have a test case testing when v is nil?

Reply:
added TestAddNilValuesToMap.